src/zarr/core/group.py

```diff
@@ -97,6 +97,24 @@ def _parse_async_node(
         raise TypeError(f"Unknown node type, got {type(node)}")
 
 
+class _MixedConsolidatedMetadataException(Exception):
```

Comment:
Let me know if this `MixConsolidatedMetadataException` stuff makes sense. The crux of the issue is that `Group.__getitem__` on a Group with consolidated metadata *can* be IO-free. To make that explicit, we define it as a plain (non-async) function.

But this `group['subgroup']['subarray']` presents a new challenge. What if `group` has consolidated metadata, but `subgroup` *doesn't*? Its consolidated metadata might be `None`, meaning we need to fall back to the non-consolidated metadata. We discover this fact in `AsyncGroup._getitem_consolidated`, which isn't an async function, so it can't call the async non-consolidated getitem.

To break through that tangle, I've added this custom exception. The expectation is that users never see it; we require (through docs & being careful?) that all the callers of `_getitem_consolidated` handle this case by catching the error and falling back if needed.

Reply:
I forgot to mention: this "mixed" case isn't going to be common. Under normal operation, where users call `consolidate_metadata` and don't mutate the group afterwards we consolidate everything and so won't hit this. I believe the only time this can happen through the public API is when users add a new group to group with consolidated metadata, and don't then re-consolidate.

Reply:
> I believe the only time this can happen through the public API is when users add a new group to group with consolidated metadata, and don't then re-consolidate.

Do we need to support this case? I would think that users could expect that, when using consolidated metadata, then metadata lookups (beyond the first) will _never_ trigger IO.

Reply:
Some discussion on this here: https://github.com/zarr-developers/zarr-python/pull/2363#issuecomment-2414491502. I think there isn't really a clear answer. I'm sympathetic to users who want to know for certain that they're done with I/O after loading a Group with consolidated metadata.

If we decide not to support this, we do still have a decision to make: when we try to index into a nested Group and reach a node with non-consolidated metadata, do we raise a `KeyError` (saying we know this key isn't there) or some public version of this MixedConsolidatedMetadataError (saying we don't know whether or not this key is there)?

Reply:
I'm not a user of consolidated metadata, so take this with a grain of salt, but I would default to being strict: if people open a group in consolidated metadata mode, then IMO the contents of the consolidated metadata should purport to provide the _complete_ model of the hierarchy, even if that model happens to be wrong :) But I'd prefer to temper my strict default with whatever expectations consolidated metadata users have.

Reply:
> but I would default to being strict

Just to be clear: what does "strict" mean to you here?

I actually am having trouble constructing a hierarchy that hits this using the public API:

```python
In [1]: import zarr

In [2]: store = zarr.storage.MemoryStore(mode="w")

In [3]: group = zarr.open_group(store=store, path="a/b/c/d")

In [4]: g = zarr.consolidate_metadata(store)

In [5]: g['a/b/c/d'].create_group("x")
Out[5]: <Group memory://4480609344/a/b/c/d/x>

In [6]: g['a/b/c/d'].metadata.consolidated_metadata
Out[6]: ConsolidatedMetadata(metadata={}, kind='inline', must_understand=False)
```

I think the semantics around what exact state in after mutating a consolidated hierarchy, but before reconsolidating, that we can do whatever we want.

Reply:
> > but I would default to being strict
> 
> Just to be clear: what does "strict" mean to you here?

for me, "strict" would mean that if a user opens a group with `use_consolidated=True, then the consolidated metadata for that group will be the single source of truth about the hierarchy, and so attempting to access a sub-node that actually exists, but isn't in that document, would raise a `KeyError`. Not sure if this is actually a reasonable position